pkg/sfu/streamallocator.go

```diff
@@ -305,60 +266,220 @@ func (s *StreamAllocator) Stop() {
+		downTrack.OnPacketSent(s.onPacketSent)
+	}
 	downTrack.AddReceiverReportListener(s.onReceiverReport)
-	downTrack.OnAvailableLayersChanged(s.onAvailableLayersChanged)
-	downTrack.OnSubscriptionChanged(s.onSubscriptionChanged)
-	downTrack.OnSubscribedLayersChanged(s.onSubscribedLayersChanged)
-	downTrack.OnPacketSent(s.onPacketSent)
+}
 
-	s.tracksMu.Lock()
-	track := newTrack(downTrack, peerID)
-	s.tracks[downTrack.ID()] = track
+func (s *StreamAllocator) RemoveTrack(downTrack *DownTrack) {
+	s.postEvent(Event{
+		Signal:    SignalRemoveTrack,
+		DownTrack: downTrack,
+	})
+}
 
-	s.tracksSorted = append(s.tracksSorted, track)
-	sort.Sort(s.tracksSorted)
+func (s *StreamAllocator) initializeEstimate() {
+	s.committedChannelCapacity = ChannelCapacityInfinity
+	s.lastCommitTime = time.Now().Add(-EstimateCommitMs)
+	s.receivedEstimate = ChannelCapacityInfinity
+	s.lastEstimateDecreaseTime = time.Now()
 
-	s.tracksMu.Unlock()
+	s.state = StateStable
+}
 
+// called when a new REMB is received
+func (s *StreamAllocator) onREMB(downTrack *DownTrack, remb *rtcp.ReceiverEstimatedMaximumBitrate) {
 	s.postEvent(Event{
-		Signal:    Signal_ADD_TRACK,
+		Signal:    SignalEstimate,
 		DownTrack: downTrack,
+		Data:      remb,
 	})
 }
 
-func (s *StreamAllocator) RemoveTrack(downTrack *DownTrack) {
-	s.tracksMu.Lock()
-	if _, ok := s.tracks[downTrack.ID()]; !ok {
-		s.tracksMu.Unlock()
+// called when a new RTCP Receiver Report is received
+func (s *StreamAllocator) onReceiverReport(downTrack *DownTrack, rr *rtcp.ReceiverReport) {
+	s.postEvent(Event{
+		Signal:    SignalReceiverReport,
+		DownTrack: downTrack,
+		Data:      rr,
+	})
+}
+
+// called when feeding track's layer availability changes
+func (s *StreamAllocator) onAvailableLayersChanged(downTrack *DownTrack) {
+	s.postEvent(Event{
+		Signal:    SignalAvailableLayersChange,
+		DownTrack: downTrack,
+	})
+}
+
+// called when subscription settings changes (muting/unmuting of track)
+func (s *StreamAllocator) onSubscriptionChanged(downTrack *DownTrack) {
+	s.postEvent(Event{
+		Signal:    SignalSubscriptionChange,
+		DownTrack: downTrack,
+	})
+}
+
+// called when subscribed layers changes (limiting max layers)
+func (s *StreamAllocator) onSubscribedLayersChanged(downTrack *DownTrack, layers VideoLayers) {
+	s.postEvent(Event{
+		Signal:    SignalSubscribedLayersChange,
+		DownTrack: downTrack,
+		Data:      layers,
+	})
+}
+
+// called when a video DownTrack sends a packet
+func (s *StreamAllocator) onPacketSent(downTrack *DownTrack, size int) {
+	s.prober.PacketSent(size)
+}
+
+// called when prober wants to send packet(s)
+func (s *StreamAllocator) onSendProbe(bytesToSend int) {
+	s.postEvent(Event{
+		Signal: SignalSendProbe,
+		Data:   bytesToSend,
+	})
+}
+
+func (s *StreamAllocator) postEvent(event Event) {
+	s.chMu.RLock()
+	defer s.chMu.RUnlock()
+
+	if !s.isRunning() {
 		return
 	}
 
-	delete(s.tracks, downTrack.ID())
+	s.eventCh <- []Event{event}
```

Comment:
No, it is one at a time. But, I was reading some blog post about golang channels and doing this :-)  I made a channel with 10 elements so that it does not block. Will revisit this in a subsequent attempt

Reply:
got it, so I would recommend creating the channel like:

```
make(chan Event, 20)
```

then send in events like s.eventCh <- event. It'll buffer 20 events.

Reply:
Oh cool, took care of it in this commit - https://github.com/livekit/livekit-server/pull/216/commits/ecdc27bbf759b154a97c09dd0ac1217524ada0b6. Thank you 🙇